src/api/source/mvt.js

```diff
@@ -68,70 +68,64 @@ export default class MVT extends Base {
-            map(propertyName => metadata.properties[propertyName].sourceName || propertyName);
-        decodingPropertyNames.forEach(propertyName => {
-            decodedProperties[propertyName] = new Float32Array(mvtLayer.length + RTT_WIDTH);
-        });
+    }
+
+    _decode(mvtLayer, metadata, mvt_extent, decodeFn) {
+        const { properties, propertyNames } = this._getPropertyNames(metadata, mvtLayer.length);
+        const geometries = decodeFn ? [] : new Float32Array(mvtLayer.length * 2);
+
         for (let i = 0; i < mvtLayer.length; i++) {
             const f = mvtLayer.feature(i);
             const geom = f.loadGeometry();
-            const mvtGeomType = f.type;
-            if (metadata.geomType === undefined) {
-                switch (mvtGeomType) {
-                    case mvtDecoderGeomTypes.point:
-                        metadata.geomType = geometryTypes.POINT;
-                        break;
-                    case mvtDecoderGeomTypes.line:
-                        metadata.geomType = geometryTypes.LINE;
-                        break;
-                    case mvtDecoderGeomTypes.polygon:
-                        metadata.geomType = geometryTypes.POLYGON;
-                        break;
-                    default:
-                        throw new Error('MVT: invalid geometry type');
-                }
-                if (metadata.geomType == geometryTypes.POINT) {
-                    points = new Float32Array(mvtLayer.length * 2);
-                }
+            if (decodeFn) {
+                const decodedPolygons = decodeFn(geom, mvt_extent);
+                geometries.push(decodedPolygons);
             }
-            if (metadata.geomType == geometryTypes.POINT) {
-                points[2 * i + 0] = 2 * (geom[0][0].x) / mvt_extent - 1.;
-                points[2 * i + 1] = 2 * (1. - (geom[0][0].y) / mvt_extent) - 1.;
-            } else if (metadata.geomType == geometryTypes.POLYGON) {
-                const decodedPolygons = decodePolygons(geom, mvt_extent);
-                featureGeometries.push(decodedPolygons);
-            } else if (metadata.geomType == geometryTypes.LINE) {
-                const decodedLines = decodeLines(geom, mvt_extent);
-                featureGeometries.push(decodedLines);
-            } else {
-                throw new Error(`Unimplemented geometry type: '${metadata.geomType}'`);
+            else {
+                geometries[2 * i + 0] = 2 * (geom[0][0].x) / mvt_extent - 1.;
+                geometries[2 * i + 1] = 2 * (1. - (geom[0][0].y) / mvt_extent) - 1.;
             }
-            decodingPropertyNames.forEach(propertyName => {
-                const propertyValue = f.properties[propertyName];
-                decodedProperties[propertyName][i] = this.decodeProperty(propertyName, propertyValue);
-            });
+            this._decodeProperties(propertyNames, properties, f, i);
```

Comment:
Also, IMO, this function should be pure and return a `properties` array instead of modifying it.

```js
properties = this._decodeProperties(propertyNames, properties, f, i);
```

Reply:
`properties` is a pre-allocated Float32Array, this cannot be pure without having an extra copy